### PR TITLE
skuba: Add new 'completion' command

### PR DESCRIPTION
**Pull request by tdaines42**
## Why is this PR needed?

As skuba is growing, new commands can be added. It's beneficial to
have a bash and zsh autocompletion commands which can be used to
generate a script which can autocomplete all skuba commands.

Fixes https://github.com/SUSE/skuba/issues/792

## What does this PR do?

Adds the commands `skuba completion bash` and `skuba completion zsh` that generate script for cli completion

## Anything else a reviewer needs to know?
Dockerfile that can be used to test:

```
FROM opensuse/leap

RUN zypper in -y git go1.12 make zsh bash-completion
ENV GOPATH=$HOME/go
ENV PATH=$GOPATH/bin:$PATH
RUN git clone https://github.com/tdaines42/skuba.git
RUN cd skuba && git checkout add-bash-completion && make install
RUN echo "source <( /go/bin/skuba completion bash | sed -n '/^#/,\$p' )" > /etc/bash_completion.d/skuba.sh
RUN skuba completion zsh | sed -n '/^#/,$p' > /etc/zsh_completion.d/_skuba
```

### Status **BEFORE** applying the patch

How can we reproduce the issue? How can we see this issue? Please provide the steps and the prove
this issue is not fixed.

### Status **AFTER** applying the patch

How can we validate this issue is fixed? Please provide the steps and the prove this issue is fixed.

## Docs

If docs need to be updated, please add a link to a PR to https://github.com/SUSE/doc-caasp.
At the time of creating the issue, this PR can be work in progress (set its title to [WIP]),
but the documentation needs to be finalized before the PR can be merged. 

# Merge restrictions

(Please do not edit this)

We are in *v4-maintenance phase*, so we will restrict what can be merged to prevent unexpected surprises:

    What can be merged (merge criteria):
        2 approvals:
            1 developer: code is fine
            1 QA: QA is fine
        there is a PR for updating documentation (or a statement that this is not needed)

### CODE REVIEW

**Comment by tdaines42** [comment]
Based on code from https://github.com/SUSE/skuba/pull/95

**Comment by tdaines42** [comment]
One thing I'm not sure about is changing the commands in the help messages to something like `skuba completion bash | tail -n +2 > "/etc/bash_completion.d/skuba.sh"` and `skuba completion zsh | tail -n +2 > "/etc/zsh_completion.d/_skuba"`

**Comment by dannysauer** [comment]
I think we should package files in the *.d directories with the skuba command, and just remove that part from the help.

We really just need to sed that first line off of the generated output, using something like `$(skuba completion bash)` in the file. Then the file contents don't have to change.  Do you know how to get another couple of files into the rpm, @tdaines?

**Comment by tdaines42** [comment]
> I think we should package files in the *.d directories with the skuba command, and just remove that part from the help.
> 
> We really just need to sed that first line off of the generated output, using something like `$(skuba completion bash)` in the file. Then the file contents don't have to change. Do you know how to get another couple of files into the rpm, @tdaines42?

I haven't had experience with packaging. I would assume there's something in IBS that could even generate the files after building skuba. 🤷‍♀

**Comment by suseclee** [comment]
> > I think we should package files in the *.d directories with the skuba command, and just remove that part from the help.
> > We really just need to sed that first line off of the generated output, using something like `$(skuba completion bash)` in the file. Then the file contents don't have to change. Do you know how to get another couple of files into the rpm, @tdaines42?
> 
> I haven't had experience with packaging. I would assume there's something in IBS that could even generate the files after building skuba.

IMO, after this PR is merged,  we need to have another issue card for changing skuba spec file to apply this.

**Comment by dannysauer** [comment]
The completion file really just needs to contain a single line in order to work.  In the case of bash, the /etc/bash_completion.d/skuba.sh file I was testing with looks like:
```bash
eval "$( ~sauer/go/bin/skuba completion bash | sed -n '/^#/,$p' )"
```
(print everything from the first comment line to the end of the file, for people who hate sed)

Though, the file should clearly use the right path for skuba. :D  That would be a nuiscance to generate in the spec file due to the quotes.  Perhaps we should just have the spec file run `skuba completion bash > skuba.sh` (probably also with the `sed` above, to clean up untagged versions) in the install stage and then copy skuba.sh into the right directory instead of the static file contents.  And repeat for zsh, of course.  That'd be pretty easy, and I think is what you're proposing, @tdaines42 

The only drawback is that it would potentially break if someone compiled a new skuba without installing the package, but do we care about that developers-only situation?  I'm thinking that we don't; people who break packages should probably expect breakage.

File: cmd/skuba/completion/bash_completions.go [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+	"k8s.io/klog"
+)
+
+// NewBashCompletion creates a `skuba completion bash` cobra command
+func NewBashCompletion() *cobra.Command {
+	return &cobra.Command{
+		Use:   "bash",
+		Short: "Generates bash completion scripts",
+		// FIXME: tail -n +2 can be dropped when stdout from main() is not polluted anymore
+		Long: `To load skuba bash completion add the following to your ~/.bashrc or ~/.profile
```
Comment (dannysauer) [inline cmd/skuba/completion/bash_completions.go]: If we generate the completion.d files, then we probably don't want people putting commands in their .profile / .bashrc.  Maybe something like:
```suggestion
		Long: `To manually load skuba bash completion, run the following command
```

File: cmd/skuba/completion/zsh_completion.go [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+// NewZshCompletion creates a `skuba completion zsh` cobra command
+func NewZshCompletion() *cobra.Command {
+	return &cobra.Command{
+		Use:   "zsh",
+		Short: "Generates zsh completion scripts",
+		// FIXME: tail -n +2 can be dropped when stdout from main() is not polluted anymore
+		Long: `To load skuba zsh completion run the following command
+
+skuba completion zsh | tail -n +2 > "${fpath[1]}/_skuba"`,
```
Comment (dannysauer) [inline cmd/skuba/completion/zsh_completion.go]: Same comments as above in the bash section, re "manually add" and using the `sed`

**Comment by dannysauer** [comment]
Should Marcos be removed from the authors in the commit so we're "all green" on the contributor check, or do we not care about that (since he did, technically, write some of the code - and was authorized to do so at the time).

File: cmd/skuba/completion/bash_completions.go [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+. <(skuba completion bash | tail -n +2)`,
```
[the remark this answers is not in the document]
Comment (tdaines42) [reply cmd/skuba/completion/bash_completions.go]: I updated the comment to match this

File: cmd/skuba/completion/bash_completions.go (hunk 1 above)
Comment (tdaines42) [reply cmd/skuba/completion/bash_completions.go]: same

File: cmd/skuba/completion/zsh_completion.go (hunk 2 above)
Comment (tdaines42) [reply cmd/skuba/completion/zsh_completion.go]: updated this one as well

**Comment by tdaines42** [comment]
I made some changes and updated the dockerfile to reflect that

**Comment by tdaines42** [comment]
> LGTM.
> 
> Though, A dockerfile is mentioned, but I don't see it in the commit - is that intentional?

It's in the comment for the PR for manual testing

**Comment by dannysauer** [comment]
I updated the skuba spec file in my skuba branch at https://build.suse.de/package/show/home:dannysauer:branches:Devel:CaaSP:4.0/skuba to build bash and zsh completion files.  When this gets merged, that spec should create two new packages - skuba-bash-completion and skuba-zsh-completion.

**Comment by dannysauer** [comment]
Binary packages at https://build.suse.de/package/binaries/home:dannysauer:branches:Devel:CaaSP:4.0/skuba/SLE_15_SP1

**Pull request merged**